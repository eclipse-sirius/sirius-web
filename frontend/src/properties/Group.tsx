/*******************************************************************************
 * Copyright (c) 2019, 2021 Obeo.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import {
  Checkbox,
  List,
  MultiSelect,
  Radio,
  Select,
  Textarea,
  Textfield,
  Widget,
  WidgetSubscription,
} from 'form/Form.types';
import { GroupProps } from 'properties/Group.types';
import { CheckboxPropertySection } from 'properties/propertysections/CheckboxPropertySection';
import { ListPropertySection } from 'properties/propertysections/ListPropertySection';
import { MultiSelectPropertySection } from 'properties/propertysections/MultiSelectPropertySection';
import { RadioPropertySection } from 'properties/propertysections/RadioPropertySection';
import { SelectPropertySection } from 'properties/propertysections/SelectPropertySection';
import { TextfieldPropertySection } from 'properties/propertysections/TextfieldPropertySection';
import React from 'react';

const useGroupStyles = makeStyles((theme) => ({
  group: {
    display: 'flex',
    flexDirection: 'column',
  },
  title: {
    whiteSpace: 'nowrap',
    overflow: 'hidden',
    textOverflow: 'ellipsis',
  },
  sections: {
    display: 'flex',
    flexDirection: 'column',
    '& > *': {
      marginBottom: theme.spacing(2),
    },
  },
}));

export const Group = ({ editingContextId, formId, group, widgetSubscriptions, readOnly }: GroupProps) => {
  const classes = useGroupStyles();

  let propertySections = group.widgets.map((widget) =>
    widgetToPropertySection(editingContextId, formId, widget, widgetSubscriptions, readOnly)
  );

  return (
    <div className={classes.group}>
      <Typography variant="subtitle1" className={classes.title} gutterBottom>
        {group.label}
      </Typography>
      <div className={classes.sections}>{propertySections}</div>
    </div>
  );
};

const isTextfield = (widget: Widget): widget is Textfield => widget.__typename === 'Textfield';
const isTextarea = (widget: Widget): widget is Textarea => widget.__typename === 'Textarea';
const isCheckbox = (widget: Widget): widget is Checkbox => widget.__typename === 'Checkbox';
const isSelect = (widget: Widget): widget is Select => widget.__typename === 'Select';
const isMultiSelect = (widget: Widget): widget is MultiSelect => widget.__typename === 'MultiSelect';
const isRadio = (widget: Widget): widget is Radio => widget.__typename === 'Radio';
const isList = (widget: Widget): widget is List => widget.__typename === 'List';

const widgetToPropertySection = (
  editingContextId: string,
  formId: string,
  widget: Widget,
  widgetSubscriptions: WidgetSubscription[],
  readOnly: boolean
) => {
  let subscribers = [];
  widgetSubscriptions
    .filter((subscription) => subscription.widgetId === widget.id)
    .forEach((subscription) => subscribers.push(...subscription.subscribers));

  let propertySection = null;
  if (isTextfield(widget) || isTextarea(widget)) {
    propertySection = (
      <TextfieldPropertySection
        editingContextId={editingContextId}
        formId={formId}
        widget={widget}
        subscribers={subscribers}
        key={widget.id}
        readOnly={readOnly}
      />
    );
  } else if (isCheckbox(widget)) {
    propertySection = (
      <CheckboxPropertySection
        editingContextId={editingContextId}
        formId={formId}
        widget={widget}
        subscribers={subscribers}
        key={widget.id}
        readOnly={readOnly}
      />
    );
  } else if (isSelect(widget)) {
    propertySection = (
      <SelectPropertySection
        editingContextId={editingContextId}
        formId={formId}
        widget={widget}
        subscribers={subscribers}
        key={widget.id}
        readOnly={readOnly}
      />
    );
  } else if (isMultiSelect(widget)) {
    propertySection = (
      <MultiSelectPropertySection
        editingContextId={editingContextId}
        formId={formId}
        widget={widget}
        subscribers={subscribers}
        key={widget.id}
        readOnly={readOnly}
      />
    );
  } else if (isRadio(widget)) {
    propertySection = (
      <RadioPropertySection
        editingContextId={editingContextId}
        formId={formId}
        widget={widget}
        subscribers={subscribers}
        key={widget.id}
        readOnly={readOnly}
      />
    );
  } else if (isList(widget)) {
    propertySection = (
      <ListPropertySection widget={widget} key={widget.id} subscribers={subscribers} readonly={readOnly} />
    );
  } else {
    console.error(`Unsupported widget type ${widget.__typename}`);
  }
  return propertySection;
};
