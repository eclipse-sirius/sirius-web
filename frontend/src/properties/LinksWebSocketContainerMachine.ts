/*******************************************************************************
 * Copyright (c) 2021 Obeo.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
import { SubscriptionResult } from '@apollo/client';
import { Form, Subscriber, WidgetSubscription } from 'form/Form.types';
import {
  GQLFormRefreshedEventPayload,
  GQLLinksEventPayload,
  GQLLinksEventSubscription,
  GQLSubscribersUpdatedEventPayload,
  GQLWidgetSubscriptionsUpdatedEventPayload,
} from 'form/FormEventFragments.types';
import { v4 as uuid } from 'uuid';
import { Selection } from 'workbench/Workbench.types';
import { assign, Machine } from 'xstate';

export interface LinksWebSocketContainerStateSchema {
  states: {
    toast: {
      states: {
        visible: {};
        hidden: {};
      };
    };
    linksWebSocketContainer: {
      states: {
        empty: {};
        unsupportedSelection: {};
        idle: {};
        ready: {};
        complete: {};
      };
    };
  };
}

export type SchemaValue = {
  toast: 'visible' | 'hidden';
  linksWebSocketContainer: 'empty' | 'unsupportedSelection' | 'idle' | 'ready' | 'complete';
};

export interface LinksWebSocketContainerContext {
  id: string;
  currentSelection: Selection | null;
  form: Form | null;
  subscribers: Subscriber[];
  widgetSubscriptions: WidgetSubscription[];
  message: string | null;
}

export type ShowToastEvent = { type: 'SHOW_TOAST'; message: string };
export type HideToastEvent = { type: 'HIDE_TOAST' };
export type SwitchSelectionEvent = { type: 'SWITCH_SELECTION'; selection: Selection; isRepresentation: boolean };
export type HandleSubscriptionResultEvent = {
  type: 'HANDLE_SUBSCRIPTION_RESULT';
  result: SubscriptionResult<GQLLinksEventSubscription>;
};
export type HandleCompleteEvent = { type: 'HANDLE_COMPLETE' };
export type LinksWebSocketContainerEvent =
  | SwitchSelectionEvent
  | HandleSubscriptionResultEvent
  | HandleCompleteEvent
  | ShowToastEvent
  | HideToastEvent;

const isFormRefreshedEventPayload = (payload: GQLLinksEventPayload): payload is GQLFormRefreshedEventPayload =>
  payload.__typename === 'FormRefreshedEventPayload';
const isSubscribersUpdatedEventPayload = (
  payload: GQLLinksEventPayload
): payload is GQLSubscribersUpdatedEventPayload => payload.__typename === 'SubscribersUpdatedEventPayload';
const isWidgetSubscriptionsUpdatedEventPayload = (
  payload: GQLLinksEventPayload
): payload is GQLWidgetSubscriptionsUpdatedEventPayload =>
  payload.__typename == 'WidgetSubscriptionsUpdatedEventPayload';

export const linksWebSocketContainerMachine = Machine<
  LinksWebSocketContainerContext,
  LinksWebSocketContainerStateSchema,
  LinksWebSocketContainerEvent
>(
  {
    type: 'parallel',
    context: {
      id: uuid(),
      currentSelection: null,
      form: null,
      subscribers: [],
      widgetSubscriptions: [],
      message: null,
    },
    states: {
      toast: {
        initial: 'hidden',
        states: {
          hidden: {
            on: {
              SHOW_TOAST: {
                target: 'visible',
                actions: 'setMessage',
              },
            },
          },
          visible: {
            on: {
              HIDE_TOAST: {
                target: 'hidden',
                actions: 'clearMessage',
              },
            },
          },
        },
      },
      linksWebSocketContainer: {
        initial: 'empty',
        states: {
          empty: {
            on: {
              SWITCH_SELECTION: [
                {
                  cond: 'isSelectionUnsupported',
                  target: 'unsupportedSelection',
                  actions: ['switchSelection', 'clearForm'],
                },
                {
                  target: 'idle',
                  actions: 'switchSelection',
                },
              ],
            },
          },
          unsupportedSelection: {
            on: {
              SWITCH_SELECTION: [
                {
                  cond: 'isSelectionUnsupported',
                  target: 'unsupportedSelection',
                  actions: ['switchSelection', 'clearForm'],
                },
                {
                  target: 'idle',
                  actions: 'switchSelection',
                },
              ],
            },
          },
          idle: {
            on: {
              SWITCH_SELECTION: [
                {
                  cond: 'isSelectionUnsupported',
                  target: 'unsupportedSelection',
                  actions: ['switchSelection', 'clearForm'],
                },
                {
                  target: 'idle',
                  actions: 'switchSelection',
                },
              ],
              HANDLE_SUBSCRIPTION_RESULT: [
                {
                  cond: 'isFormRefreshedEventPayload',
                  target: 'ready',
                  actions: 'handleSubscriptionResult',
                },
                {
                  target: 'idle',
                  actions: 'handleSubscriptionResult',
                },
              ],
            },
          },
          ready: {
            on: {
              SWITCH_SELECTION: [
                {
                  cond: 'isSelectionUnsupported',
                  target: 'unsupportedSelection',
                  actions: ['switchSelection', 'clearForm'],
                },
                {
                  target: 'idle',
                  actions: 'switchSelection',
                },
              ],
              HANDLE_SUBSCRIPTION_RESULT: {
                target: 'ready',
                actions: 'handleSubscriptionResult',
              },
              HANDLE_COMPLETE: {
                target: 'complete',
              },
            },
          },
          complete: {
            on: {
              SWITCH_SELECTION: [
                {
                  cond: 'isSelectionUnsupported',
                  target: 'unsupportedSelection',
                  actions: ['switchSelection', 'clearForm'],
                },
                {
                  target: 'idle',
                  actions: 'switchSelection',
                },
              ],
            },
          },
        },
      },
    },
  },
  {
    guards: {
      isFormRefreshedEventPayload: (_, event) => {
        const { result } = event as HandleSubscriptionResultEvent;
        const { data } = result;
        return isFormRefreshedEventPayload(data.linksEvent);
      },
      isSelectionUnsupported: (_, event) => {
        const { selection, isRepresentation } = event as SwitchSelectionEvent;
        return !selection || isRepresentation || selection.kind === 'Unknown' || selection.kind === 'Document';
      },
    },
    actions: {
      switchSelection: assign((_, event) => {
        const { selection } = event as SwitchSelectionEvent;
        return { id: uuid(), currentSelection: selection };
      }),
      clearForm: assign((_, event) => {
        return { form: null };
      }),
      handleSubscriptionResult: assign((_, event) => {
        const { result } = event as HandleSubscriptionResultEvent;
        const { data } = result;
        if (isFormRefreshedEventPayload(data.linksEvent)) {
          const { form } = data.linksEvent;
          return { form };
        } else if (isSubscribersUpdatedEventPayload(data.linksEvent)) {
          const { subscribers } = data.linksEvent;
          return { subscribers };
        } else if (isWidgetSubscriptionsUpdatedEventPayload(data.linksEvent)) {
          const { widgetSubscriptions } = data.linksEvent;
          return { widgetSubscriptions };
        }
        return {};
      }),
      setMessage: assign((_, event) => {
        const { message } = event as ShowToastEvent;
        return { message };
      }),
      clearMessage: assign((_) => {
        return { message: null };
      }),
    },
  }
);
