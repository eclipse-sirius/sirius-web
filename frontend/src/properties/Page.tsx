/*******************************************************************************
 * Copyright (c) 2019, 2021 Obeo.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
import { makeStyles } from '@material-ui/core/styles';
import { Group } from 'properties/Group';
import { PageProps } from 'properties/Page.types';
import React from 'react';

const usePageStyles = makeStyles((theme) => ({
  page: {
    display: 'flex',
    flexDirection: 'column',
    '& > *': {
      marginBottom: theme.spacing(2),
    },
  },
}));

export const Page = ({ editingContextId, formId, page, widgetSubscriptions, setSelection, readOnly }: PageProps) => {
  const classes = usePageStyles();
  return (
    <div className={classes.page}>
      {page.groups.map((group) => {
        return (
          <Group
            editingContextId={editingContextId}
            formId={formId}
            group={group}
            widgetSubscriptions={widgetSubscriptions}
            key={group.id}
            setSelection={setSelection}
            readOnly={readOnly}
          />
        );
      })}
    </div>
  );
};
