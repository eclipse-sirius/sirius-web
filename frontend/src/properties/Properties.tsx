/*******************************************************************************
 * Copyright (c) 2019, 2021 Obeo.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
import Avatar from '@material-ui/core/Avatar';
import { makeStyles } from '@material-ui/core/styles';
import Toolbar from '@material-ui/core/Toolbar';
import Tooltip from '@material-ui/core/Tooltip';
import Typography from '@material-ui/core/Typography';
import { Page } from 'properties/Page';
import { PageList } from 'properties/pagelist/PageList';
import { FormProps } from 'properties/Properties.types';
import React from 'react';

const usePropertiesStyles = makeStyles((theme) => ({
  properties: {
    display: 'flex',
    flexDirection: 'column',
    paddingLeft: theme.spacing(1),
    paddingRight: theme.spacing(1),
  },
  content: {
    overflowX: 'hidden',
    overflowY: 'auto',
  },
  subscribers: {
    marginLeft: 'auto',
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    '& > *': {
      marginLeft: theme.spacing(0.5),
      marginRight: theme.spacing(0.5),
    },
  },
  avatar: {
    fontSize: '1rem',
    width: theme.spacing(3),
    height: theme.spacing(3),
  },
  container: {
    display: 'grid',
    gridTemplateRows: '1fr',
    gridTemplateColumns: 'min-content 1fr',
  },
}));

export const Properties = ({ editingContextId, form, subscribers, widgetSubscriptions, readOnly }: FormProps) => {
  const classes = usePropertiesStyles();
  const { id, label, pages } = form;

  let content;
  if (pages.length > 1) {
    content = (
      <div className={classes.container}>
        <PageList pages={pages} />
        <Page
          editingContextId={editingContextId}
          formId={id}
          page={pages[0]}
          widgetSubscriptions={widgetSubscriptions}
          readOnly={readOnly}
        />
      </div>
    );
  } else if (pages.length === 1) {
    content = (
      <Page
        editingContextId={editingContextId}
        formId={id}
        page={pages[0]}
        widgetSubscriptions={widgetSubscriptions}
        readOnly={readOnly}
      />
    );
  }
  return (
    <div data-testid="properties" className={classes.properties}>
      <Toolbar variant="dense" disableGutters>
        <Typography variant="h6" noWrap>
          {label}
        </Typography>
        <div className={classes.subscribers}>
          {subscribers.map((subscriber) => (
            <Tooltip title={subscriber.username} arrow key={subscriber.username}>
              <Avatar classes={{ root: classes.avatar }}>{subscriber.username.substring(0, 1).toUpperCase()}</Avatar>
            </Tooltip>
          ))}
        </div>
      </Toolbar>
      <div className={classes.content}>{content}</div>
    </div>
  );
};
