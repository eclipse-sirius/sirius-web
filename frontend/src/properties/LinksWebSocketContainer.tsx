/*******************************************************************************
 * Copyright (c) 2021 Obeo.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
import { useSubscription } from '@apollo/client';
import IconButton from '@material-ui/core/IconButton';
import Snackbar from '@material-ui/core/Snackbar';
import makeStyles from '@material-ui/core/styles/makeStyles';
import Typography from '@material-ui/core/Typography';
import CloseIcon from '@material-ui/icons/Close';
import { useMachine } from '@xstate/react';
import {
  formRefreshedEventPayloadFragment,
  subscribersUpdatedEventPayloadFragment,
  widgetSubscriptionsUpdatedEventPayloadFragment,
} from 'form/FormEventFragments';
import { GQLLinksEventSubscription } from 'form/FormEventFragments.types';
import gql from 'graphql-tag';
import {
  HandleCompleteEvent,
  HandleSubscriptionResultEvent,
  HideToastEvent,
  LinksWebSocketContainerContext,
  LinksWebSocketContainerEvent,
  linksWebSocketContainerMachine,
  SchemaValue,
  ShowToastEvent,
  SwitchSelectionEvent,
} from 'properties/LinksWebSocketContainerMachine';
import { Properties } from 'properties/Properties';
import React, { useContext, useEffect } from 'react';
import { RepresentationContext } from 'workbench/RepresentationContext';
import { LinksWebSocketContainerProps } from './LinksWebSocketContainer.types';

const linksEventSubscription = gql`
  subscription linksEvent($input: LinksEventInput!) {
    linksEvent(input: $input) {
      __typename
      ... on SubscribersUpdatedEventPayload {
        ...subscribersUpdatedEventPayloadFragment
      }
      ... on WidgetSubscriptionsUpdatedEventPayload {
        ...widgetSubscriptionsUpdatedEventPayloadFragment
      }
      ... on FormRefreshedEventPayload {
        ...formRefreshedEventPayloadFragment
      }
    }
  }
  ${subscribersUpdatedEventPayloadFragment}
  ${widgetSubscriptionsUpdatedEventPayloadFragment}
  ${formRefreshedEventPayloadFragment}
`;

const usePropertiesWebSocketContainerStyles = makeStyles((theme) => ({
  idle: {
    padding: theme.spacing(1),
  },
}));

/**
 * Connect the Properties component to the GraphQL API over Web Socket.
 */
export const LinksWebSocketContainer = ({ editingContextId, selection }: LinksWebSocketContainerProps) => {
  const classes = usePropertiesWebSocketContainerStyles();
  const [{ value, context }, dispatch] = useMachine<LinksWebSocketContainerContext, LinksWebSocketContainerEvent>(
    linksWebSocketContainerMachine
  );
  const { toast, linksWebSocketContainer } = value as SchemaValue;
  const { id, currentSelection, form, subscribers, widgetSubscriptions, message } = context;
  const { registry } = useContext(RepresentationContext);

  /**
   * Displays an other form if the selection indicates that we should display another properties view.
   */
  useEffect(() => {
    if (currentSelection?.id !== selection?.id) {
      const isRepresentation = registry.isRepresentation(selection.kind);
      const switchSelectionEvent: SwitchSelectionEvent = { type: 'SWITCH_SELECTION', selection, isRepresentation };
      dispatch(switchSelectionEvent);
    }
  }, [currentSelection, registry, selection, dispatch]);

  const { error } = useSubscription<GQLLinksEventSubscription>(linksEventSubscription, {
    variables: {
      input: {
        id,
        editingContextId,
        objectId: currentSelection?.id,
      },
    },
    fetchPolicy: 'no-cache',
    skip: linksWebSocketContainer === 'empty' || linksWebSocketContainer === 'unsupportedSelection',
    onSubscriptionData: ({ subscriptionData }) => {
      const handleDataEvent: HandleSubscriptionResultEvent = {
        type: 'HANDLE_SUBSCRIPTION_RESULT',
        result: subscriptionData,
      };
      dispatch(handleDataEvent);
    },
    onSubscriptionComplete: () => {
      const completeEvent: HandleCompleteEvent = { type: 'HANDLE_COMPLETE' };
      dispatch(completeEvent);
    },
  });

  useEffect(() => {
    if (error) {
      const { message } = error;
      const showToastEvent: ShowToastEvent = { type: 'SHOW_TOAST', message };
      dispatch(showToastEvent);
    }
  }, [error, dispatch]);

  let content = null;
  if (!selection || linksWebSocketContainer === 'unsupportedSelection') {
    content = (
      <div className={classes.idle}>
        <Typography variant="subtitle2">No object selected</Typography>
      </div>
    );
  }
  if ((linksWebSocketContainer === 'idle' && form) || linksWebSocketContainer === 'ready') {
    content = (
      <Properties
        editingContextId={editingContextId}
        form={form}
        subscribers={subscribers}
        widgetSubscriptions={widgetSubscriptions}
        readOnly={true}
      />
    );
  }
  return (
    <>
      {content}
      <Snackbar
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'right',
        }}
        open={toast === 'visible'}
        autoHideDuration={3000}
        onClose={() => dispatch({ type: 'HIDE_TOAST' } as HideToastEvent)}
        message={message}
        action={
          <IconButton
            size="small"
            aria-label="close"
            color="inherit"
            onClick={() => dispatch({ type: 'HIDE_TOAST' } as HideToastEvent)}>
            <CloseIcon fontSize="small" />
          </IconButton>
        }
        data-testid="error"
      />
    </>
  );
};
