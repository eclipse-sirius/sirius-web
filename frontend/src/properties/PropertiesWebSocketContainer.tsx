/*******************************************************************************
 * Copyright (c) 2019, 2021 Obeo.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
import { useSubscription } from '@apollo/client';
import IconButton from '@material-ui/core/IconButton';
import Snackbar from '@material-ui/core/Snackbar';
import makeStyles from '@material-ui/core/styles/makeStyles';
import Typography from '@material-ui/core/Typography';
import CloseIcon from '@material-ui/icons/Close';
import { useMachine } from '@xstate/react';
import {
  formRefreshedEventPayloadFragment,
  subscribersUpdatedEventPayloadFragment,
  widgetSubscriptionsUpdatedEventPayloadFragment,
} from 'form/FormEventFragments';
import { GQLPropertiesEventSubscription } from 'form/FormEventFragments.types';
import gql from 'graphql-tag';
import { Properties } from 'properties/Properties';
import { PropertiesWebSocketContainerProps } from 'properties/PropertiesWebSocketContainer.types';
import {
  HandleCompleteEvent,
  HandleSubscriptionResultEvent,
  HideToastEvent,
  PropertiesWebSocketContainerContext,
  PropertiesWebSocketContainerEvent,
  propertiesWebSocketContainerMachine,
  SchemaValue,
  ShowToastEvent,
  SwitchSelectionEvent,
} from 'properties/PropertiesWebSocketContainerMachine';
import React, { useContext, useEffect } from 'react';
import { RepresentationContext } from 'workbench/RepresentationContext';

const propertiesEventSubscription = gql`
  subscription propertiesEvent($input: PropertiesEventInput!) {
    propertiesEvent(input: $input) {
      __typename
      ... on SubscribersUpdatedEventPayload {
        ...subscribersUpdatedEventPayloadFragment
      }
      ... on WidgetSubscriptionsUpdatedEventPayload {
        ...widgetSubscriptionsUpdatedEventPayloadFragment
      }
      ... on FormRefreshedEventPayload {
        ...formRefreshedEventPayloadFragment
      }
    }
  }
  ${subscribersUpdatedEventPayloadFragment}
  ${widgetSubscriptionsUpdatedEventPayloadFragment}
  ${formRefreshedEventPayloadFragment}
`;

const usePropertiesWebSocketContainerStyles = makeStyles((theme) => ({
  idle: {
    padding: theme.spacing(1),
  },
}));

/**
 * Connect the Properties component to the GraphQL API over Web Socket.
 */
export const PropertiesWebSocketContainer = ({
  editingContextId,
  selection,
  readOnly,
}: PropertiesWebSocketContainerProps) => {
  const classes = usePropertiesWebSocketContainerStyles();
  const [{ value, context }, dispatch] = useMachine<
    PropertiesWebSocketContainerContext,
    PropertiesWebSocketContainerEvent
  >(propertiesWebSocketContainerMachine);
  const { toast, propertiesWebSocketContainer } = value as SchemaValue;
  const { id, currentSelection, form, subscribers, widgetSubscriptions, message } = context;
  const { registry } = useContext(RepresentationContext);

  /**
   * Displays an other form if the selection indicates that we should display another properties view.
   */
  useEffect(() => {
    if (currentSelection?.id !== selection?.id) {
      const isRepresentation = registry.isRepresentation(selection.kind);
      const switchSelectionEvent: SwitchSelectionEvent = { type: 'SWITCH_SELECTION', selection, isRepresentation };
      dispatch(switchSelectionEvent);
    }
  }, [currentSelection, registry, selection, dispatch]);

  const { error } = useSubscription<GQLPropertiesEventSubscription>(propertiesEventSubscription, {
    variables: {
      input: {
        id,
        editingContextId,
        objectId: currentSelection?.id,
      },
    },
    fetchPolicy: 'no-cache',
    skip: propertiesWebSocketContainer === 'empty' || propertiesWebSocketContainer === 'unsupportedSelection',
    onSubscriptionData: ({ subscriptionData }) => {
      const handleDataEvent: HandleSubscriptionResultEvent = {
        type: 'HANDLE_SUBSCRIPTION_RESULT',
        result: subscriptionData,
      };
      dispatch(handleDataEvent);
    },
    onSubscriptionComplete: () => {
      const completeEvent: HandleCompleteEvent = { type: 'HANDLE_COMPLETE' };
      dispatch(completeEvent);
    },
  });

  useEffect(() => {
    if (error) {
      const { message } = error;
      const showToastEvent: ShowToastEvent = { type: 'SHOW_TOAST', message };
      dispatch(showToastEvent);
    }
  }, [error, dispatch]);

  let content = null;
  if (!selection || propertiesWebSocketContainer === 'unsupportedSelection') {
    content = (
      <div className={classes.idle}>
        <Typography variant="subtitle2">No object selected</Typography>
      </div>
    );
  }
  if ((propertiesWebSocketContainer === 'idle' && form) || propertiesWebSocketContainer === 'ready') {
    content = (
      <Properties
        editingContextId={editingContextId}
        form={form}
        subscribers={subscribers}
        widgetSubscriptions={widgetSubscriptions}
        readOnly={readOnly}
      />
    );
  }
  return (
    <>
      {content}
      <Snackbar
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'right',
        }}
        open={toast === 'visible'}
        autoHideDuration={3000}
        onClose={() => dispatch({ type: 'HIDE_TOAST' } as HideToastEvent)}
        message={message}
        action={
          <IconButton
            size="small"
            aria-label="close"
            color="inherit"
            onClick={() => dispatch({ type: 'HIDE_TOAST' } as HideToastEvent)}>
            <CloseIcon fontSize="small" />
          </IconButton>
        }
        data-testid="error"
      />
    </>
  );
};
