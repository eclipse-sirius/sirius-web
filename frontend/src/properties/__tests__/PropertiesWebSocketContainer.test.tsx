/*******************************************************************************
 * Copyright (c) 2022 Obeo.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { act, render, screen, waitFor } from '@testing-library/react';
import {
  GQLCheckbox,
  GQLForm,
  GQLFormRefreshedEventPayload,
  GQLGroup,
  GQLPage,
  GQLPropertiesEventSubscription,
  GQLPropertiesEventVariables,
  GQLTextfield,
} from 'form/FormEventFragments.types';
import { propertiesEventSubscription, PropertiesWebSocketContainer } from 'properties/PropertiesWebSocketContainer';
import React from 'react';
import { Selection } from 'workbench/Workbench.types';

jest.mock('uuid', () => ({ v4: () => '48be95fc-3422-45d3-b1f9-d590e847e9e1' }));

const defaultCheckbox: GQLCheckbox = {
  __typename: 'Checkbox',
  id: 'checkboxId',
  label: 'CheckboxLabel',
  diagnostics: [],
  booleanValue: false,
};

const defaultTextField: GQLTextfield = {
  __typename: 'Textfield',
  id: 'textfieldId',
  label: 'Name:',
  stringValue: 'Composite Processor',
  diagnostics: [],
};

const defaultGroup: GQLGroup = {
  id: 'groupId',
  label: 'groupeLabel',
  widgets: [defaultCheckbox, defaultTextField],
};

const defaultPage: GQLPage = {
  id: 'pageId',
  label: 'pageLabel',
  groups: [defaultGroup],
};

const defaultGQLForm: GQLForm = {
  id: 'formId',
  metadata: undefined,
  pages: [defaultPage],
};

const propertiesEventSubscriptionSuccessPayload: GQLFormRefreshedEventPayload = {
  __typename: 'FormRefreshedEventPayload',
  id: 'payloadId',
  form: defaultGQLForm,
};

const propertiesEventFormRefreshedSubscription: GQLPropertiesEventSubscription = {
  propertiesEvent: propertiesEventSubscriptionSuccessPayload,
};

const propertiesEventSubscriptionVariables: GQLPropertiesEventVariables = {
  input: {
    id: '48be95fc-3422-45d3-b1f9-d590e847e9e1',
    editingContextId: 'editingContextId',
    objectIds: ['selectionId'],
  },
};

test('should render the form', async () => {
  let propertiesEventSubscriptionCalled = false;
  const propertiesEventSubscriptionMock: MockedResponse<Record<string, any>> = {
    request: {
      query: propertiesEventSubscription,
      variables: propertiesEventSubscriptionVariables,
    },
    result: () => {
      propertiesEventSubscriptionCalled = true;
      return { data: propertiesEventFormRefreshedSubscription };
    },
  };

  const selection: Selection = {
    entries: [
      {
        id: 'selectionId',
        label: 'selectionLabel',
        kind: 'selectionKind',
      },
    ],
  };
  const mocks = [propertiesEventSubscriptionMock];
  const { container } = render(
    <MockedProvider mocks={mocks}>
      <PropertiesWebSocketContainer
        editingContextId="editingContextId"
        readOnly={false}
        selection={selection}
        setSelection={function (selection: Selection): void {}}
      />
    </MockedProvider>
  );

  await act(async () => {
    await waitFor(() => {
      //The problem is that the mock sends the subscription termination signal
      //and I did not find how to mock it.
      //This test works before the signal is sent, but is rerun when the termination signal is received
      //since the component is rerendered.
      expect(propertiesEventSubscriptionCalled).toBeTruthy();
      expect(container).toMatchSnapshot();

      const textField: HTMLInputElement = screen.getByTestId('Name:').querySelector('input');
      expect(textField.value).toBe('Composite Processor');

      const checkBox: HTMLInputElement = screen.getByTestId('CheckboxLabel').querySelector('input');
      expect(checkBox.checked).toBe(false);
    });
  });
});
