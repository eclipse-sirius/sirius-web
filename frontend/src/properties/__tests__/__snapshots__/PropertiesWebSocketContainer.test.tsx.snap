// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the form 1`] = `
<div>
  <div
    class="makeStyles-idle-1"
  >
    <h6
      class="MuiTypography-root MuiTypography-subtitle2"
    >
      No object selected 
      complete
    </h6>
  </div>
</div>
`;

exports[`should render the form 2`] = `
<div>
  <div
    class="makeStyles-idle-1"
  >
    <h6
      class="MuiTypography-root MuiTypography-subtitle2"
    >
      No object selected 
      complete
    </h6>
  </div>
</div>
`;

exports[`should render the form 3`] = `
<div>
  <div
    class="makeStyles-idle-1"
  >
    <h6
      class="MuiTypography-root MuiTypography-subtitle2"
    >
      No object selected 
      complete
    </h6>
  </div>
</div>
`;
